articles/ros2_dds_security.md

```diff
@@ -0,0 +1,220 @@
+---
+layout: default
+title: ROS 2 DDS-Security integration
+permalink: articles/ros2_dds_security.html
+abstract: >
+  Robotics is full of experimentation: evaluating different hardware and software, pushing ahead with what works, and culling what doesn't.
+  ROS was designed to be flexible to enable this experimentation; to allow existing components to easily be combined with new ones or swapped with others.
+  In ROS 1, this flexibility was valued above all else, at the cost of security.
+  By virtue of being designed on top of DDS, ROS 2 is able to retain that flexibility while obtaining the ability to be secured by properly utilizing the DDS-Security specification.
+  This article describes how ROS 2 integrates with DDS-Security.
+author:  >
+  [Kyle Fazzari](https://github.com/kyrofa)
+
+published: true
+categories: Security
+---
+
+{:toc}
+
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+
+# DDS-Security overview
+
+The [DDS-Security specification][dds_security] expands upon the [DDS specification][dds], adding security enhancements by defining a Service Plugin Interface (SPI) architecture, a set of builtin implementations of the SPIs, and the security model enforced by the SPIs. Specifically, there are five SPIs defined:
+
+- **Authentication**: Verify the identity of a given domain participant
+- **Access control**: Enforce restrictions on the DDS-related operations that can be performed by an authenticated domain participant
+- **Cryptographic**: Takes care of all encryption, signing, and hashing operations required
+- **Logging**: Provides the ability to audit DDS-Security-related events
+- **Data tagging**: Provides the ability to add tags to data samples
+
+ROS 2's security features currently utilize only the first three of these, as not all DDS implementations support the **Logging** or **Data Tagging** plugins.
+Let's delve a little further into those first three plugins.
+
+
+## Authentication
+
+The **Authentication** plugin (see section 8.3 of the [DDS-Security spec][dds_security]) is central to the entire SPI architecture, as it provides the concept of a confirmed identity without which further enforcement would be impossible (e.g. it would be awfully hard to make sure a given ROS node could only access specific topics if it was impossible to securely determine which node it was).
+
+The SPI architecture allows for a number of potential authentication schemes, but ROS 2 uses the builtin authentication plugin (called "DDS:Auth:PKI-DH", see section 9.3 of the [DDS-Security spec][dds_security]), which uses the proven Public Key Infrastructure (PKI).
```

Comment:
I disagree. What I'm trying to say here is that DDS-Security doesn't mandate how authentication is actually supposed to work, and ROS 2 happens to use a specific implementation that works THIS way. It doesn't have to, it could have used anything, and we might change it at some point. I could just as happily change this to "While the SPI architecture allows for a number of potential authentication schemes, ROS 2 uses [...]" but the "while" seems to be serving the same purpose, there. Do you like that better?

Reply:
Thanks for the detailed answer.
I don't feel strongly about the wording and lgtm as is :+1:

---

As this is reflecting a restrictive choice we could consider providing a rationale regarding the choice of using builtin plugins (for Auth and Access Control); though the rationale would be limited to the fact that this is the only approach described in details in the spec and that the mandatory support for its implementation and interoperability between vendors made it the most sensible option.
+1 for getting this merged with or without a rationale on the plugins choice.

Reply:
> we could consider providing a rationale regarding the choice of using builtin plugins

Oh yeah, great idea! Done.